system/lib/wasmfs/wasmfs.cpp

```diff
@@ -61,31 +62,103 @@ __wasi_errno_t __wasi_fd_write(__wasi_fd_t fd,
                                const __wasi_ciovec_t* iovs,
                                size_t iovs_len,
                                __wasi_size_t* nwritten) {
+  if (iovs_len < 0) {
+    return __WASI_ERRNO_INVAL;
+  }
+
   auto openFile = FileTable::get()[fd];
 
   if (!openFile) {
     return __WASI_ERRNO_BADF;
   }
 
-  auto file = openFile.locked().getFile()->dynCast<DataFile>();
+  auto lockedOpenFile = openFile.locked();
+  auto* file = lockedOpenFile.getFile()->dynCast<DataFile>();
 
   // If file is nullptr, then the file was not a DataFile.
+  // TODO: change to add support for symlinks.
   if (!file) {
     return __WASI_ERRNO_ISDIR;
   }
 
   auto lockedFile = file->locked();
 
-  __wasi_size_t num = 0;
+  off_t offset = lockedOpenFile.position();
   for (size_t i = 0; i < iovs_len; i++) {
     const uint8_t* buf = iovs[i].buf;
-    __wasi_size_t len = iovs[i].buf_len;
+    size_t len = iovs[i].buf_len;
+
+    // Check if the sum of the buf_len values overflows a uint64_t (file size).
+    if (addWillOverFlow(wasmfs_offset_t(offset), wasmfs_offset_t(len))) {
+      return __WASI_ERRNO_FBIG;
+    }
+
+    // Check if buf_len specifies a positive length buffer but buf is a
+    // null pointer
+    if (!buf && len > 0) {
+      return __WASI_ERRNO_INVAL;
+    }
+
+    lockedFile.write(buf, len, offset);
+    offset += len;
+  }
+
+  *nwritten = offset - lockedOpenFile.position();
+  lockedOpenFile.position() = offset;
+
+  return __WASI_ERRNO_SUCCESS;
+}
+
+__wasi_errno_t __wasi_fd_read(__wasi_fd_t fd,
+                              const __wasi_iovec_t* iovs,
+                              size_t iovs_len,
+                              __wasi_size_t* nread) {
+  if (iovs_len < 0) {
+    return __WASI_ERRNO_INVAL;
+  }
 
-    lockedFile.write(buf, len);
-    num += len;
+  auto openFile = FileTable::get()[fd];
+
+  if (!openFile) {
+    return __WASI_ERRNO_BADF;
+  }
+
+  auto lockedOpenFile = openFile.locked();
+  auto* file = lockedOpenFile.getFile()->dynCast<DataFile>();
+
+  // If file is nullptr, then the file was not a DataFile.
+  // TODO: change to add support for symlinks.
+  if (!file) {
+    return __WASI_ERRNO_ISDIR;
   }
-  *nwritten = num;
 
+  auto lockedFile = file->locked();
+
+  off_t offset = lockedOpenFile.position();
+  size_t size = lockedFile.size();
+  for (size_t i = 0; i < iovs_len; i++) {
+    // Check if offset has exceeded size of file data.
+    ssize_t dataLeft = size - offset;
+    if (dataLeft <= 0) {
+      break;
```

Comment:
If an attempt is made to read more than is available, do we not need to return EOF or such?